staking/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+            + (exec::block_timestamp() - self.produced_time) as u128
+                / self.distribution_time as u128
+    }
+
+    /// Updates the reward produced so far and calculates tokens per stake
+    fn update_reward(&mut self) {
+        let reward_produced_at_now = self.produced();
+
+        debug!("reward_produced_at_now: {}", reward_produced_at_now);
+
+        if reward_produced_at_now > self.reward_produced {
+            let produced_new = reward_produced_at_now - self.reward_produced;
+
+            if self.total_staked > 0 {
+                self.tokens_per_stake = self
+                    .tokens_per_stake
+                    .saturating_add(produced_new * (1 << 20) / self.total_staked);
+            }
+
+            self.reward_produced = self.reward_produced.saturating_add(produced_new);
+        }
+    }
+
+    /// Calculates the reward of the staker that is currently avaiable
+    fn calc_reward(&mut self) -> u128 {
+        if let Some(staker) = self.stakers.get(&msg::source()) {
+            return (staker.balance * self.tokens_per_stake) / (1 << 20) + staker.reward_allowed
+                - staker.reward_debt
+                - staker.distributed;
+        }
+
+        panic!("calc_reward(): Staker {:?} not found", msg::source());
+    }
+
+    /// Stakes the tokens
+    async fn stake(&mut self, amount: u128) {
+        if amount > 0 {
+            self.update_reward();
+
+            self.stakers
+                .entry(msg::source())
+                .and_modify(|stake| {
+                    stake.reward_debt = stake
+                        .reward_debt
+                        .saturating_add((amount * self.tokens_per_stake) / (1 << 20));
+                    stake.balance = stake.balance.saturating_add(amount);
+                })
+                .or_insert(Staker {
+                    reward_debt: (amount * self.tokens_per_stake) / (1 << 20),
+                    balance: amount,
+                    ..Default::default()
+                });
+
+            self.total_staked = self.total_staked.saturating_add(amount);
+
+            let token_address = self.staking_token_address;
+
+            self.transfer_tokens(&token_address, &msg::source(), &exec::program_id(), amount)
+                .await;
+
+            msg::reply(StakingEvent::StakeAccepted(amount), 0).unwrap();
+        } else {
+            panic!("enter(): amount: {}", amount);
+        }
+    }
+
+    ///Sends reward to the staker
+    async fn send_reward(&mut self) {
+        self.update_reward();
+        let reward = self.calc_reward();
+
+        if reward > 0 {
+            self.stakers
+                .entry(msg::source())
+                .and_modify(|stake| stake.distributed = stake.distributed.saturating_add(reward));
+
+            let mut token_address = self.staking_token_address;
+
+            if self.reward_token_address.is_some() {
+                token_address = self.reward_token_address.unwrap();
+            }
```

Comment:
Usually, we use `if let` here instead of `unwrap()`.
```suggestion
            if self.reward_token_address.is_some() {
                token_address = self.reward_token_address.unwrap();
            }
```
```suggestion
            if let Some(address) = self.reward_token_address {
                token_address = address;
            }
```

Reply:
Fixed